### PR TITLE
Count leading zero bits in one CPU tick

### PR DESCRIPTION
**Pull request by hnarkaytis**
Count of leading bits was implemented as a CPU instructing 6 years ago. It doesn't make sense to have custom code for this.
https://en.wikipedia.org/wiki/Bit_Manipulation_Instruction_Sets#ABM_(Advanced_Bit_Manipulation)

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@hnarkaytis

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Review by remilapeyre — COMMENTED** [review]
Is this supported on all architecture that Python support (https://www.python.org/download/other/)?
What about CPU that don't have this instruction or where this instruction is slower?
Also this would need to be compatible with all compilers supported by CPython.

**Comment by mdickinson** [comment]
Python runs on many different platforms and architectures. It's written (mostly) in standard C (specifically, C89, but with some C99 features permitted). `__builtin_clz` is not part of standard C.

It's okay to have platform-specific code when there's a good reason for it, but there will still need to be fallback code for platforms that don't have `__builtin_clz` available: we can't simply assume that it'll be available on any non-Windows platform.

File: Python/pymath.c [hunk 1]
```diff
@@ -80,17 +80,13 @@ round(double x)
...
 
 unsigned int _Py_bit_length(unsigned long d) {
-   unsigned int d_bits = 0;
-   while (d >= 32) {
-       d_bits += 6;
-       d >>= 6;
-   }
-   d_bits += BitLengthTable[d];
-   return d_bits;
+   return d ? CHAR_BIT * sizeof (d) - __builtin_clz (d) : 0;
```
Comment (mdickinson) [inline Python/pymath.c]: This is the wrong function. You want `__builtin_clzl`, not `__builtin_clz`. But in any case, this isn't going to work unless there's also a fallback for platforms that don't provide `__builtin_clz`.

**Comment by mdickinson** [comment]
@hnarkaytis Thanks for the contribution. If you want to pursue this further, there are a couple of steps you need to take:

- Please could you open a corresponding issue on the bug tracker (https://bugs.python.org/)? At the moment, we don't use GitHub issues, so bugs.python.org is where any related discussion should take place.
- Please sign the CLA agreement (see the automated messages above) if you haven't done so already.

See the [Python Developer's Guide](https://devguide.python.org) for more detailed information and guidelines on contributing to core Python.

**Comment by hnarkaytis** [comment]
@mdickinson let me take care of this.
I will extend configure for __builtin_clzl autodetection and will add fallback implementation for outdated compilers.

HHN.

> @hnarkaytis Thanks for the contribution. If you want to pursue this further, there are a couple of steps you need to take:
> 
> * Please could you open a corresponding issue on the bug tracker (https://bugs.python.org/)? At the moment, we don't use GitHub issues, so bugs.python.org is where any related discussion should take place.
> * Please sign the CLA agreement (see the automated messages above) if you haven't done so already.
> 
> See the [Python Developer's Guide](https://devguide.python.org) for more detailed information and guidelines on contributing to core Python.

**Comment by ned-deily** [comment]
@hnarkaytis, sorry but if you want anyone to review your suggestions, you need to follow our process as @mdickinson explained.  And you need to redo this PR so it is against the current top of the master branch only, such that it only shows the actual changes you propose, not all the merges in 372 files.

**Comment by hnarkaytis** [comment]
> @hnarkaytis, sorry but if you want anyone to review your suggestions, you need to follow our process as @mdickinson explained. And you need to redo this PR so it is against the current top of the master branch only, such that it only shows the actual changes you propose, not all the merges in 372 files.

Yeah - I know. Just pushed code first. Will get the rest shortly. I converted pull request to draft, so it should not trigger anyone in review list.

**Pull request closed without merging**